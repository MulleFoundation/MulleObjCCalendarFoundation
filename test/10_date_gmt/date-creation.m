#ifdef __MULLE_OBJC__
# import <MulleObjCCalendarFoundation/MulleObjCCalendarFoundation.h>
#else
# import <Foundation/Foundation.h>
#endif


//
// the values in stdout were produced by macOS
//

struct date
{
   int   day;
   int   month;
   int   year;

   int   hour;
   int   minute;
   int   second;
} dates[] =
{
   {  1,  1, 1970,  12,  0,  0 },
   { 18,  4, 1848,  18, 48,  0 },
   {  1,  1, 2000,  12,  0,  0 },
   { 31, 12, 2000,  23, 59, 59 },

   // schaltjahr
   { 28, 2, 2020,  23, 59, 59 },
   { 29, 2, 2020,   0,  0,  0 },
   { 29, 2, 2020,  23, 59, 59 },
   { 30, 2, 2020,   0,  0,  0 },
   {  1, 3, 2020,   0,  0,  0 },

   // kein schaltjahr
   { 28, 2, 2100,  23, 59, 59 },
   { 29, 2, 2100,  23, 59, 59 },

   // sommerzeit
   { 31, 3, 2019,  1, 59, 59 },
   { 31, 3, 2019,  2,  0,  0 },
   { 31, 3, 2019,  2, 59, 59 },
   { 31, 3, 2019,  3,  0,  0 }
};


int   main( void)
{
   NSCalendar        *calendar;
   NSDateComponents  *components;
   NSDate            *date;
   struct date       *p;
   struct date       *sentinel;

   calendar   = [NSCalendar calendarWithIdentifier:NSGregorianCalendar];
   [calendar setTimeZone:[NSTimeZone timeZoneWithAbbreviation:@"GMT"]];

   p        = &dates[ 0];
   sentinel = &p[ sizeof( dates) / sizeof( dates[ 0])];
   for( ;p < sentinel; ++p)
   {
      components = [[NSDateComponents new] autorelease];

      if( p->day)
         [components setDay:p->day];
      if( p->month)
         [components setMonth:p->month];
      if( p->year)
         [components setYear:p->year];
      if( p->hour >= 0)
         [components setHour:p->hour];
      if( p->minute >= 0)
         [components setMinute:p->minute];
      if( p->second >= 0)
         [components setSecond:p->second];

      date = [calendar dateFromComponents:components];
      printf( "%d.%d.%d %d:%02d:%02d = %.1f\n",
            p->day, p->month, p->year,
            p->hour, p->minute, p->second,
            [date timeIntervalSinceReferenceDate]);
   }

   return( 0);
}