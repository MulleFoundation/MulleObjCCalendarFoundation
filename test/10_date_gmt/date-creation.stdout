1.1.1970 12:00:00 = -978264000.0
18.4.1848 18:48:00 = -4818892320.0
1.1.2000 12:00:00 = -31579200.0
31.12.2000 23:59:59 = -1.0
28.2.2020 23:59:59 = 604627199.0
29.2.2020 0:00:00 = 604627200.0
29.2.2020 23:59:59 = 604713599.0
30.2.2020 0:00:00 = 604713600.0
1.3.2020 0:00:00 = 604713600.0
28.2.2100 23:59:59 = 3129235199.0
29.2.2100 23:59:59 = 3129321599.0
31.3.2019 1:59:59 = 575690399.0
31.3.2019 2:00:00 = 575690400.0
31.3.2019 2:59:59 = 575693999.0
31.3.2019 3:00:00 = 575694000.0
